### ISSUE TITLE
Extend CleanVision to run on video data

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanvision#215]
Some discussion of this here:  https://github.com/cleanlab/cleanvision/discussions/214

Making CleanVision runnable on video would be super useful to many folks!

Should be fairly easy:

- just extract every k-th frame from the video,
- run cleanvision on those,
- aggregate the results across frames per video

**Comment by LemurPwned** [comment]
I'd be super interested in this -- might come in handy when sampling for annotations etc. Potentially doing some on-the-fly deduplication using the perceptual hash to avoid sampling visually identical scenes, before passing in to do further analysis, could improve efficiency of the process.

**Comment by jwmueller** [comment]
@LemurPwned This would be an awesome contribution!  Nobody is working on this actively AFAIK. 

@lafemme https://github.com/lafemmephile 
was interested in trying CleanVision for video and could test out your PR code too!

**Comment by LemurPwned** [comment]
@jwmueller Great, I started playing around with the idea and for simplicity sake I created a separate repo for the sampling part: https://github.com/LemurPwned/video-sampler.
It seems that depending on the settings a significant reduction can take place. It's also pretty damn quick!

 I'll try to integrate the sampling with the clean vision api in a PR soon.

**Comment by jwmueller** [comment]
awesome super excited about this!!

**Comment by lafemmephile** [comment]
> I'd be super interested in this -- might come in handy when sampling for annotations etc. Potentially doing some on-the-fly deduplication using the perceptual hash to avoid sampling visually identical scenes, before passing in to do further analysis, could improve efficiency of the process.

I like the *deduplication* feature ... I saw the code you wrote ... looks very interesting. Seems like you've done some good work there, I want to study it more in depth.

**Comment by LemurPwned** [comment]
@lafemmephile thanks, just to be clear, you want to pick up from here?

**Comment by lafemmephile** [comment]
> @lafemmephile thanks, just to be clear, you want to pick up from here?

@LemurPwned  Yes. I'm going to take a stab at assembling this `Videolab` class/module, which is much easier than the work you've already done (again thank you so much for developing the video sampling solution).

With the work you've already done (i.e. the *frame sampling*), it seems we are tantalizingly close to a *minimum viable product* (i.e. something approximating `cleanvision.imagelab.Imagelab` but for *video data* aka `cleanvision.videolab.Videolab`).

The approach I am roughly considering:
+ mirror package/module/class structure of `cleanvision.imagelab.Imagelab` with `cleanvision.videolab.Videolab`
+ subclass `Imagelab`:
   ```python
    from cleanvision.imagelab import Imagelab

    class Videolab(Imagelab):
       # .....
    ```
 + override methods where *video sampling* must occur first

Any comments/thoughts/concerns pleas share them ... because otherwise this is roughly what I see as the way to implement this *video feature* (which is really just an entirely new sub-module with a new `Videolab` class).

**Comment by LemurPwned** [comment]
@lafemmephile thanks for clarifying :) -- I'm fine with this, let me know if you need any help. 
As to the outline you mentioned, makes sense -- the trickest part is probably avoiding reimplementation of the features from Imagelab.

**Comment by lafemmephile** [comment]
> @lafemmephile thanks for clarifying :) -- I'm fine with this, let me know if you need any help. As to the outline you mentioned, makes sense -- the trickest part is probably avoiding reimplementation of the features from Imagelab.

@LemurPwned Yes I can see what you mean already ... it could be a little tricky with trying to *extend* the `Imagelab` class but without *reinventing the wheel* ... we'll see how it goes.

**Comment by lafemmephile** [comment]
> @jwmueller Great, I started playing around with the idea and for simplicity sake I created a separate repo for the sampling part: https://github.com/LemurPwned/video-sampler. It seems that depending on the settings a significant reduction can take place. It's also pretty damn quick!
> 
> I'll try to integrate the sampling with the clean vision api in a PR soon.

@LemurPwned So I've had a good look at the overall *design* of the entire `cleanvision` library, and I now have a sense of what the *approach* [I outlined](https://github.com/cleanlab/cleanvision/issues/215#issuecomment-1737560147) will require. But I am curious what your approach was? How would you go about *integrating* the `video_sampler` code with the *cleanvision API*?

**Comment by LemurPwned** [comment]
@lafemmephile I'm not a maintainer for this project, so I'm not sure to what extent my insight is useful here. But, I was essentially thinking of something similar to [spacy's pipeline](https://github.com/explosion/spaCy/blob/4ec41e98f622302b70af037dfb5b43c2967533b9/spacy/language.py#L1016). A simple, minimal object which takes in the config for each step in the pipeline and then executes them in sequence.
So, create a small video sampler object as a separate API, define its config, then write a simple pipeline executor instance that runs the sampler, saves the output and points image lab at it. In principle, prepping the config input correctly (as a kwargs dict or something to that extent) would decouple those objects from most changes in `Imagelab`'s object api. 
And, it allows you to add more steps before/after if needed.

**Comment by sanjanag** [comment]
Hi @LemurPwned ! The video sampler works great. Could you explain in more details about the approach you're suggesting for integrating video sampler in cleanvision? More specifically a pseudocode for what the classes would look like?

@lafemmephile We could start with a VideoLab class, that'd take care of sampling frames from videos, saving them and running cleanvision on those frames. I'd suggest to also think about a few finer points while working on this integration
- What issues should VideoLab check for? For example, odd_size would be rendered useless in the case of a video since all images will be the same size.
- The VideoLab code is decoupled from the Imagelab code as much as possible to avoid any breaking changes in Imagelab.
- What are the preprocessing steps needed for VideoLab? For instance, sampling frames is a preprocessing step. 
- What's the structure of the data object that will carried through VideoLab. For example in Imagelab we use the [Dataset](https://github.com/cleanlab/cleanvision/blob/main/src/cleanvision/dataset/base_dataset.py) object to store all information related to dataset. Can we use the same or need to make a new one?
- How would the visualizations of results look like for Videolab?
- How would one detect duplicate videos? We can start with image property issues to make it simple.

@lafemmephile Feel free to start more discussions on these points. I think we could start with a layout of the code first and then start filling in details for specific methods.

**Comment by lafemmephile** [comment]
> @lafemmephile I'm not a maintainer for this project, so I'm not sure to what extent my insight is useful here. But, I was essentially thinking of something similar to [spacy's pipeline](https://github.com/explosion/spaCy/blob/4ec41e98f622302b70af037dfb5b43c2967533b9/spacy/language.py#L1016). A simple, minimal object which takes in the config for each step in the pipeline and then executes them in sequence. So, I'd create a small video sampler object as a separate API, define its config, then write a simple pipeline executor instance that runs the sampler, saves the output and points image lab at it. In principle, prepping the config input correctly (as a kwargs dict or something to that extent) would decouple those objects from most changes in `Imagelab`'s object api. And, it allows you to add more steps before/after if needed.

@LemurPwned Here's my understanding of your approach (correct me if I am wrong):
```console
/path/to/video/files ---> VideoSampler ---> /path/to/video/frame/samples ---> Imagelab
```
To clarify, in your approach you see that the *video sampling* would occur first, creating a directory of samples, with each *video file* having its own sub directory of *video frame samples* in the same manner that `video_sampler` stores frame samples. From here, the parent directory holding all the samples can be passed to `Imagelab`.

This is more or less the approach I envisioned, but how accurate is my understanding? Or did you have something different than what I described?

**Comment by lafemmephile** [comment]
> Hi @LemurPwned ! The video sampler works great. Could you explain in more details about the approach you're suggesting for integrating video sampler in cleanvision? More specifically a pseudocode for what the classes would look like?
> 
> @lafemmephile We could start with a VideoLab class, that'd take care of sampling frames from videos, saving them and running cleanvision on those frames. I'd suggest to also think about a few finer points while working on this extension.
> 
>     * What issues should VideoLab check for? For example, odd_size would be rendered useless in the case of a video since all images will be the same size.
> 
>     * The VideoLab code is decoupled from the Imagelab code as much as possible to avoid any breaking changes in Imagelab.
> 
>     * What are the preprocessing steps needed for VideoLab? For instance, sampling frames is a preprocessing step.
> 
>     * What's the structure of the data object that will carried through VideoLab. For example in Imagelab we use the [Dataset](https://github.com/cleanlab/cleanvision/blob/main/src/cleanvision/dataset/base_dataset.py) object to store all information related to dataset. Can we use the same or need to make a new one?
> 
>     * How would the visualizations of results look like for Videolab?
> 
>     * How would one detect duplicate videos? We can start with image property issues to make it simple.
> 
> 
> @lafemmephile Feel free to start more discussions on these points. I think we could start with a layout of the code first and then start filling in details for specific methods.

@sanjanag Excellent, I have already been thinking about this and have at least addressed one of these points:
+ "What's the structure of the data object that will carried through VideoLab. For example in Imagelab we use the [Dataset](https://github.com/cleanlab/cleanvision/blob/main/src/cleanvision/dataset/base_dataset.py) object to store all information related to dataset. Can we use the same or need to make a new one?"


Should be able to keep the fundamental `Dataset` class, but add some additional `data_type` information for switching between *different types* of data (i.e. `image`, `video`, `audio`, `text`, etc ...).

I address this approach here: https://github.com/lafemmephile/cleanvision/issues/2#issuecomment-1739591855

**Comment by lafemmephile** [comment]
>  I think we could start with a layout of the code first and then start filling in details for specific methods.

@sanjanag what should this layout look like? Are we talking *pseudocode* or just some simple flowchart?

**Comment by LemurPwned** [comment]
@lafemmephile your understanding is correct.

@sanjanag I was just thinking about a minimal implementation. Here's a more detailed outline.
1. Modify [FSDataset dataset](https://github.com/cleanlab/cleanvision/blob/main/src/cleanvision/dataset/fsspec_dataset.py) slightly by allowing a switch from `IMAGE_FILE_EXTENSIONS` to `VIDEO_FILE_EXTENSIONS`. I think other datasets classes don't really apply since we will be intentionally subsampling. 
2. Create an object which you would call `VideoLab` that accepts said `Dataset` object.
3. I'm not sure to what extent a concept of an issue applies to a video. For sure, in big, mixed datasets we may require a removal of overly dark, or blurry frames. Here lies the problem. If you decide to include more complex issue filters in the video, then running `ImageLab`-like afterwards may become redundant. Instead, I can see how one may be more inclined to keep `ImageLab` as your primary object where all the heavy duty analytics happens and `VideoLab` does only minimal video deduplication such that the downstream tasks are not too time consuming. On the other hand you may opt for putting some more filtering into the `VideoLab` object and fine-tune/optimise it for the video use case. I don't know which one is the right answer at this point -- but there are definitely some gains in optimising for the video sampling performance.

As to the API outline, I was thinking something relatively simple as this (the pipeline is very crude and not abstracted here):
```
from tempfile import TemporaryDirectory
VIDEO_FILE_EXTENSIONS = ["*.mp4", "*.avi", "*.mkv", "*.mov", "*.webm"]

class VideoPipeline:
    def __init__(self, input_folder, output_folder) -> None:
        # overwrite default allowed extensions
        dataset = FSDataset(input_folder, allowed_extensions=VIDEO_FILE_EXTENSIONS)
        # create a temporary folder to store the output
        self.output_folder = output_folder
        self.tmp_out = TemporaryDirectory()
        self.video_lab = VideoLab(dataset=dataset, output_dataset=self.tmp_out)
        self.image_lab = ImageLab(dataset=self.tmp_out)

    def run(self, sampling_config, issue_config, n_jobs: int):
        self.video_lab.sample(**sampling_config, n_jobs=n_jobs)
        self.image_lab.find_issues(**issue_config, n_jobs=n_jobs)
        # save images to the output folder, maybe move images?
        self.image_lab.save(self.output_folder)
        self.tmp_out.cleanup()
```
The `VideoLab` would look very similar to what you've seen in the `video-sampler` example repo.

**Comment by smttsp** [comment]
Hey everyone, 

I have some experience with videos, so I may have some suggestions that might be helpful: 

1. you guys can use I-frames instead of just randomly sampling some frames. Typically there is 1 I-frame per second, they are a bit less noisy compared to P or B-frames. One way to do this is using ffmpeg, something like this should work:

```
pip install ffmpeg-python
```

and 

```
import ffmpeg

def extract_i_frames(input_video_path, output_directory):
    # Create the output directory if it doesn't exist
    if not os.path.exists(output_directory):
        os.makedirs(output_directory)

    # FFmpeg command to extract I-frames
    ffmpeg.input(input_video_path).output(os.path.join(output_directory, 'frame%d.jpg'), vf='select=eq(pict_type\\,I)', vsync=0).run()

    print('I-frames extracted successfully.')

# Example usage
input_video_path = 'input_video.mp4'  # Replace with the path to your input video file
output_directory = 'output_frames'  # Replace with the desired output directory

extract_i_frames(input_video_path, output_directory)
```

2. I would write an algorithm in a way that it is easily extensible to other 3D volumes, such as CT-scan, MRI. CT-scan is basically a couple of hundred layers of images. So, once `VideoPipeline` is completed, the algorithm should pretty much work for CT scan with some small wrapper function. 

3. not sure how useful this would be for you guys but I implemented an algorithm recently which takes a video and using a YOLO model, it finds temporal inconsistencies in a video: https://github.com/smttsp/temporal_consistency_odt. I am not an expert in this field, so I cannot think of a way to integrate something like that to this repo, but just an idea. 

Let me know if I can help any other way

**Comment by LemurPwned** [comment]
@smttsp in the (video-sampler)[https://github.com/LemurPwned/video-sampler] example repo I'm using keyframe decoding with pyav. That's what ffmpeg uses under the hood (not the Python version, but the C version).  This gives you a programmatic access to frames as they are being decoded (and all of their metadata such as motion vectors) which I think is more flexible than using ffmpeg bindings.

Ad. 3 I took a look at your repo and it's super interesting! I'm contemplating extending the example repo with an arbitrary filter that could operate on K accumulated frames (so it's easily extendable) which would make it possible to operate with  the combination of detection+tracking, like you did, on the fly.

**Comment by smttsp** [comment]
@LemurPwned, cool, both sounds good.

**Comment by lafemmephile** [comment]
@smttsp Very interesting work with the *temporal consistency* ... I want to look over your work more in-depth.

To @LemurPwned's **3rd** point:
> I'm not sure to what extent a concept of an issue applies to a video. For sure, in big, mixed datasets we may require a removal of overly dark, or blurry frames. Here lies the problem. If you decide to include more complex issue filters in the video, then running ImageLab-like afterwards may become redundant. Instead, I can see how one may be more inclined to keep ImageLab as your primary object where all the heavy duty analytics happens and VideoLab does only minimal video deduplication such that the downstream tasks are not too time consuming. On the other hand you may opt for putting some more filtering into the VideoLab object and fine-tune/optimise it for the video use case. I don't know which one is the right answer at this point -- but there are definitely some gains in optimising for the video sampling performance.

I see now that answering **this question** (i.e. what *video issues* to look for) fundamentally determines the approach to designing, and implementing the `Videolab` class.

@sanjanag @smttsp @LemurPwned Maybe we can begin to discuss what issues make sense to start with ...

**Comment by sanjanag** [comment]
Hi @lafemmephile ! The idea of this Github issue to extend what cleanvision detects to videos. Hence, we strictly want to focus on extending cleanvision issues for videos. 

`What issues (if any) must be developed exclusively for Videolab (i.e. new issues only used for video data)?` needs more brainstorming and we must take it up in a separate issue.

The class for sampling frames from videos is a good starting point and it seems like you also have a decent idea of what would go in VideoLab. I'd suggest that at this point you can create a PR, which would first detect image property issues in issues in videos like dark, light, etc.

**Comment by sanjanag** [comment]
Hi @smttsp ! Great to see your inputs on this issue. https://github.com/smttsp/temporal_consistency_odt looks great. Our team recently added [support for object detection](https://docs.cleanlab.ai/stable/tutorials/object_detection.html) in [cleanlab](https://github.com/cleanlab/cleanlab) package.  Seems like we could use your inputs there.

**Comment by smttsp** [comment]
Hey @lafemmephile, 

> @smttsp Very interesting work with the temporal consistency ... I want to look over your work more in-depth.

Thank you! LMK if I can help in any way or if you can think of any feature that can be directly or indirectly used here. We can also think of incorporating the visualizations/exports. 

> Maybe we can begin to discuss what video issues make sense to start with 

One of the major considerations is whether we will evaluate videos or individual frames, i.e., what will be displayed in the tool.

I think frames would make more sense. Suppose that you have two 1-minute videos with 10 seconds overlap. So, the videos are near duplicates. But each video has 50 seconds of unique content. If we categorize them as near duplicate, then you need to discard one of them which doesn't make sense because you are discarding one of the 50 seconds.

If frames are in question, you can discard one of the 10 seconds and you are fine. 

But the problem is then, what if the video is from a static video where there are several near duplicate images? 

**HERE IS QUITE GOOD I THINK:**

If we export the frames in a smarter way, i.e., only export the unique frames, then we can have intra-video (frame-wise) uniqueness.

Then near duplicate frames can only occur between videos.

simple pseudocode for a video frame extraction
```
frame_hash_set = set()
for frame in selected_frames:
    if hash(frame) not in frame_hash_set:
         frame_hash_set.add(hash(frame))
         export frame
    else: continue
```

**Comment by smttsp** [comment]
Forgot to answer these:

> 1. How many issues currently used in Imagelab can be used directly without modification in Videolab?
2. How many issues currently used in Imagelab can be used with some modification in Videolab?
3. What issues (if any) must be developed exclusively for Videolab (i.e. new issues only used for video data)?

The above suggestion is basically converting a video into a set of unique-frames which would enable to use all the features of imagelab. The only issue I see here is that visualization of videos should be a bit different from images. Not sure how to do that in an elegant way but we can discuss that later on

**Comment by smttsp** [comment]
> Hi @smttsp ! Great to see your inputs on this issue. https://github.com/smttsp/temporal_consistency_odt looks great. Our team recently added [support for object detection](https://docs.cleanlab.ai/stable/tutorials/object_detection.html) in [cleanlab](https://github.com/cleanlab/cleanlab) package. Seems like we could use your inputs there.

@sanjanag, of course, I would be happy to help!

**Comment by lafemmephile** [comment]
> The only issue I see here is that the visualization of videos should be a bit different from images. Not sure how to do that in an elegant way but we can discuss that later on

@smttsp You are correct about the *video data visualization* ... but as @sanjanag said:
> The idea of this Github issue to extend what cleanvision detects to videos. Hence, we strictly want to focus on extending cleanvision issues for videos.

That may be outside of the scope of this *current issue*, but either way hold onto that thought, because eventually it will need to be considered.

> One of the major considerations is whether we will evaluate videos or individual frames.

> I think frames would make more sense. Suppose that you have two 1-minute videos with 10 seconds overlap. So, the videos are near duplicates. But each video has 50 seconds of unique content. If we categorize the videos as near duplicates, then you need to discard one of them which doesn't make sense because we are discarding one of the 50 seconds (unique content of one of the videos).

> If frames are in question, you can discard one of the 10 seconds and you are fine.

> But the problem is then, what if we have a static video where there are several near duplicate images?

> I like the following solution a lot:

> If we export the frames in a smarter way, i.e., only export the unique frames, we can have intra-video (frame-wise) uniqueness.

> Then near duplicate frames can only occur between two different videos.

@smttsp THIS is very good, what you are thinking about, because you are really thinking about the *big picture problem* that needs to be considered and solved wisely/intelligently:
> How to apply Imagelab issues to Videolab data? How do we evolve from finding issues in image files, to finding issues in video files?

Although *this issue* only seeks to *extend* the current code, the questions you are asking are really where the potential POWER of the `Videolab` class can manifest (and is extremely relevant to my own interests). 

> The above suggestion is basically converting a video into a set of unique frames which would enable to use all the features of imagelab.

@smttsp I am curious how your *frame sampling* approach differs from @LemurPwned's [VideoSampler](https://github.com/LemurPwned/video-sampler)? Again, lots of really good ideas being put forth, I cannot wait to mature this current *extended solution* to the point where we can look at some of these more advanced *video file diffs* or *video file introspection* techniques. Very exciting stuff.

**Comment by lafemmephile** [comment]
> Hi @lafemmephile ! The idea of this Github issue to extend what cleanvision detects to videos. Hence, we strictly want to focus on extending cleanvision issues for videos.
> 
> `What issues (if any) must be developed exclusively for Videolab (i.e. new issues only used for video data)?` needs more brainstorming and we must take it up in a separate issue.
> 
> The class for sampling frames from videos is a good starting point and it seems like you also have a decent idea of what would go in VideoLab. I'd suggest that at this point you can create a PR, which would first detect image property issues in issues in videos like dark, light, etc.

@sanjanag Understood. The scope is now even more clear and defined. I will begin developing the *layout* of the *implementation* of the `Videolab` class, while carefully considering the points you mentioned:

+ What issues should VideoLab check for? For example, odd_size would be rendered useless in the case of a video since all  images will be the same size.
+ The VideoLab code is decoupled from the Imagelab code as much as possible to avoid any breaking changes in Imagelab.
+ What are the preprocessing steps needed for VideoLab? For instance, sampling frames is a preprocessing step.
+ What's the structure of the data object that will carried through VideoLab. For example in Imagelab we use the [Dataset](https://github.com/cleanlab/cleanvision/blob/main/src/cleanvision/dataset/base_dataset.py) object to store all information related to dataset. Can we use the same or need to make a new one?
+ How would the visualizations of results look like for Videolab?
+ How would one detect duplicate videos? We can start with image property issues to make it simple.

And staying within the scope of *existing issues*:

+ How many issues currently used in Imagelab can be used directly without modification in Videolab?
+ How many issues currently used in Imagelab can be used with some modification in Videolab?
+ ~~What issues (if any) must be developed exclusively for Videolab (i.e. new issues only used for video data)?~~

**Comment by smttsp** [comment]
> @smttsp I am curious how your frame sampling approach differs from @LemurPwned's [VideoSampler](https://github.com/LemurPwned/video-sampler)? Again, lots of really good ideas being put forth, I cannot wait to mature this current extended solution to the point where we can look at some of these more advanced video file diffs or video file introspection techniques. Very exciting stuff.

Actually, just read the VideoSampler code more thoroughly and it seems that it is already using p-hash. great job @LemurPwned  👏 👏 👏

**Comment by smttsp** [comment]
> How many issues currently used in Imagelab can be used directly without modification in Videolab?

I think all issues can be used if the goal is to build the tool on a frame-by-frame basis. But then, what is the point of calling it `VideoLab`? 

> How many issues currently used in Imagelab can be used with some modification in Videolab?

If the goal is to compare videos, then all of the issues require some extra work. For example, exact duplication of videos is when all the frames of two videos are exactly the same, and `near duplicate` is when at least `k` of `n` (above a threshold) frames are near duplicates. 

Light, blurry, dark, and low information are all gonna be based on some threshold. We may have a few frames to be one of those but the rest of the video might be perfect. 

I think VideoLab should find both `frame-related` and `video-related` issues. Everything ImageLab does should be frame-by-frame basis. Along with those, we can come up with a couple of other issues specifically for videos. A few things I can think of are:

> What issues (if any) must be developed exclusively for Videolab (i.e. new issues only used for video data)

- `video-type`: static video vs non-static video or other types we can think of. Based on this, we can tell a video has low information vs high information.
- `light/dark/low information`: if the entire video consists of such frames. There can be a little bit different naming such as `low-information-video`, `saturated/overexposed-video`, `dark-video` etc.
- `poor quality or highly-compressed` videos
- `very-long` videos (e.g. a 90-min football match, how much information will those frames (90*60*30 frames) add to model training). Say that we are getting 1 frame per second, it is 5400 frames.

**Comment by LemurPwned** [comment]
@smttsp 
I think it's a very valid distinction -- what I don't have a clear idea on is how to do that effectively over a large dataset. 
For instance, taking the low-information case as an example, I can image we arrive at a 2D histogram wanting to compute the video entropy  and, say, solve some constrained tree optimisation problem where you sample X frames with some minimal temporal distance apart (to avoid sampling from clusters of temporally close high entropy pictures). But, this _probably_ decomposes into sampling with the same constraint first and then running image lab with low info issue? At the same time, it's probably more work to define that as a video issue.

**Comment by jwmueller** [comment]
@lafemmephile There's a lot of nice ideas discussed here, but I still think your first PR should keep things simple. Future PRs can extend the set of video issues that can be detected further, but IMO it's best to focus on getting that first PR in now. 

I think the first PR can stick with the original strategy I outlined:
1. just extract every k-th frame from the video,
2. run cleanvision on all those images,
3. aggregate the results across the frames extracted per video, into scores per video.

Note the complexity is in Step 3. We need to define aggregation functions to get issue scores per video out of the individual frames' issue scores. For example: the dark-score for a video might be the 0.95-percentile of the frames' dark scores (so that we only say a video is dark if most of its frames are dark).

As to which issue types make sense here, I'd first stick with just a couple in the first PR. We don't need to support every CleanVision issue type in the 1st PR for `Videolab`. I'd rather see just a couple issue types supported well, and then launch `Videolab`, rather than spending a long time to refine all of the issue types for `Videolab` simultaneously. It is better to add support for the rest of the CleanVision issue types in later PRs, and then add support for video-specific issue types in subsequent PRs after that.

**Comment by lafemmephile** [comment]
>I think the first PR can stick with the original strategy I outlined:
> + just extract every k-th frame from the video,
> + run cleanvision on all those images,
> + aggregate the results across the frames extracted per video, into scores per video.

@jwmueller I agree.

> As to which issue types make sense here, I'd first stick with just a couple in the first PR (eg. light/dark, blurry). We don't need to support every CleanVision issue type in the 1st PR for Videolab. I'd rather see just a couple issue types supported well, and then launch Videolab, rather than spending a long time to refine all of the issue types for Videolab simultaneously. It is better to add support for the rest of the CleanVision issue types in later PRs, and then add support for video-specific issue types in subsequent PRs after that.

@jwmueller I'm close to having this *MVP* (minimum viable product) that you suggest, and I understand your position on this. There is a lot of good discussion from @smttsp and @LemurPwned, they have a solid grasp of the *advanced video issues* that would eventually be worth developing.

We just need to store these good ideas, and focus on the *MVP* that @jwmueller suggested from the beginning. Once that is accomplished, more *issues* (GitHub issues to be precise) can be opened to *improve* the *MVP*.
